### ISSUE TITLE
EVP_PKEY_fromdata/OSSL_PARAM_BN requires data in both big and little endian

### ISSUE THREAD

**Issue by kroeckx** [openssl/openssl#12388]
The EVP_PKEY_fromdata() manpage has this in it:
```
#ifndef B_ENDIAN
# error "We haven't prepared little endian arrays"
#endif
```

I think the API is wrong if the application needs to know the endianness and nees to have an array in both big and little endian.

(It should also say big, not little.)

**Comment by romen** [comment]
IIRC in the past months we had a discussion on this matter: pinging @levitte @paulidale @slontis

**Comment by levitte** [comment]
This has been a discussion from the very early start.  In the first iteration of `OSSL_PARAM` definition, back when we were designing it, I had INTEGER data looking like the DER encoded ASN.1 INTEGER (without the DER tag and length), i.e. MSB first and an extra 0x00 at the start of a positive number that had the highest bit set.

With that form, *all* native C integer types would have to be copied to a buffer on LSB first systems, i.e. we would do the classic network\<-\>host number conversion that we've learned from dealing with diverse network packets.  BIGNUMs would be copied to buffers as well, using classic `BN_bn2bin()`, which just so happens to organise the data MSB first.

This met very hard pushback at the time, and the argument that won it was that if the `OSSL_PARAM` INTEGER type was organised in native form, you could have the `OSSL_PARAM` *data* pointer point directly at the native C integer that contains the data instead of having to perform the network\<-\>host number transform.  `BIGNUM`s still needed to be copied to a buffer, and `BN_bn2native()` was invented for that very purpose.  Still, we gained something with that modification.

Now, we have *a few* examples where we don't start with a `BIGNUM`, but rather create the buffer it would have been copied to directly and verbatim.  It could be wise to modify them to actually show `BIGNUM` use, instead of cheating like this.  After all, that's exactly what we do in for calls like `RSA_set0_key()`, so we should probably do the same for `EVP_PKEY_fromdata()`.

-----

It's been suggested once that there should be a separate `BIGNUM` `OSSL_PARAM` type.  I disagree with that idea completely, it doesn't relate anything about the `OSSL_PARAM` data organisation...  all it does is to transmit *an OpenSSL implementation detail*, that there are certain integers that OpenSSL implements as `BIGNUM`s and others not.  A provider may very well make a different choice, and having this different treatment for some integers may increase their burden unnecessarily.

**Comment by levitte** [comment]
Do note that we have since added quite a bit of functionality that makes life easier.  Dealing directly with `OSSL_PARAM` details is very raw, and it's possible that the `EVP_PKEY_fromdata()` isn't the best place to demonstrate such raw form.  The example can be rewritten so it doesn't ruffle feathers quite so much :wink:

**Comment by kroeckx** [comment]
> BIGNUMs would be copied to buffers as well, using classic `BN_bn2bin()`, which just so happens to organise the data MSB first.

BIGNUMs are little endian on a little endian machine, but middle
endian on a big endian machine. That is, the lowest word contains
the least significate word in it in native format.

**Comment by levitte** [comment]
> BIGNUMs are little endian on a little endian machine, but middle endian on a big endian machine. That is, the lowest word contains the least significate word in it in native format.

That's internal OpenSSL implementation details.  Other bignum libraries are organised differently.  Do ponder that  a provider should be obligated to use *our* BIGNUM implementation, let alone link with libcrypto, or even be written specifically in C.  `OSSL_PARAM` is meant to be agnostic of OpenSSL implementation details.  If that wasn't the case, we could just as well have attached BIGNUMs directly, and let the provider writers that dares deal directly with the BIGNUM limbs as they see fit.

**Comment by levitte** [comment]
One way to fix the direct issue is #12389, i.e. a more flexible example.
